2-WebApp-graph-user/2-5-HybridFlow/README.md

```diff
@@ -0,0 +1,771 @@
+            // is passed along in the 'Microsoft.Identity.Hybrid.Authentication' value.
+            //
+            // https://docs.microsoft.com/en-us/dotnet/api/microsoft.identity.client.authenticationresult?view=azure-dotnet
+            var authResult = await confidentialClientService.GetAuthenticationResultAsync(options.Scope, context.ProtocolMessage.Code, codeVerifier);
+
+            context.Request.HttpContext.Session.SetString("Microsoft.Identity.Hybrid.Authentication", authResult.SpaAuthCode);
+
+            context.HandleCodeRedemption(authResult.AccessToken, authResult.IdToken);
+        };
+
+        // Even though the ASP.NET Core middleware handles the call to the Microsoft Identity Platform logout path
+        // automatically this application creates a ConfidentialClientApplication in order to handle the auth token
+        // flow which also has it's own cache. In order to ensure that cached tokens for users are removed this handler
+        // is activated after a user is redirected to the application from the Microsoft Identity logout endpoint.
+        //
+        // You can find more information here:
+        // https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/tree/master/1-WebApp-OIDC/1-6-SignOut
+        options.Events.OnRedirectToIdentityProviderForSignOut = async context =>
+        {
+            var oid = context.HttpContext.User.GetObjectId();
+            var tid = context.HttpContext.User.GetTenantId();
+
+            if (!string.IsNullOrEmpty(oid) && !string.IsNullOrEmpty(tid))
+            {
+                await confidentialClientService.RemoveAccount($"{oid}.{tid}");
+            }
+        };
+
+        builder.Configuration.Bind("AzureAd", options);
+    });
+```
+
+The rest of the file configures a basic **ASP.NET Core** app to use **Razor** pages, **Controllers** and a few other configurations.
+
+```CSharp
+builder.Services.AddSession();
+builder.Services.AddControllersWithViews(options =>
+{
+    var policy = new AuthorizationPolicyBuilder()
+        .RequireAuthenticatedUser()
+        .Build();
+
+    options.Filters.Add(new AuthorizeFilter(policy));
+}).AddMicrosoftIdentityUI();
+
+builder.Services.AddRazorPages();
+
+var app = builder.Build();
+
+app.UseSession();
+
+// Configure the HTTP request pipeline.
+if (!app.Environment.IsDevelopment())
+{
+    app.UseExceptionHandler("/Error");
+    // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
+    app.UseHsts();
+}
+
+app.UseHttpsRedirection();
+app.UseStaticFiles();
+
+app.UseRouting();
+
+app.UseAuthentication();
+app.UseAuthorization();
+
+app.UseEndpoints(endpoints =>
+{
+    endpoints.MapControllerRoute(
+        name: "default",
+        pattern: "{controller=Home}/{action=Index}/{id?}");
+    endpoints.MapRazorPages();
+});
+
+app.Run();
+```
+
+## Server-side Token Redemption
```

Comment:
>Server-side Token Redemption [](http://example.com/codeflow?start=3&length=28)

Well its technically right, but can potentially confuse as we use "redemption" in SPA later. I'd suggest replacing with "Acquring the AT and authz code via hybrid SPA flow" or something like that :)